### DISCUSSION TITLE
How to generate  standalone project lock file based on single  lock file in workspace root.

### DISCUSSION THREAD

**Discussion by EscapeB** [pnpm/pnpm#4247] [Q&A]
I am using pnpm to manage packages for our big monorepo project (200+ projects and 100+ developers). 
We facing problem in resolve git conflicts in single lock file, and changes in single lock file from one project may also affect another project in monorepo. 

We plan to turn off [shared-workspace-lockfile](https://pnpm.io/workspaces#shared-workspace-lockfile) and generate lock file for each project in monorepo.

But here comes the question that we need to generate project lock file based on previous single lock file because there are some compatibility issues in some packages we used. So we have to use the lock version in previous single lock file.

Is there any recommended way to do this work ?

******
I have wrote an script to parse previous single lock file and  get the dependency tree for one projects recursively.

This script generate new pnpm-lock.yaml in each projects. 

New pnpm-lock.yaml works with command `pnmp install`but fails with `pnpm install --fix-lockfile`

**Comment by zkochan** [comment]
You can probably use this package to create the new lockfiles: https://github.com/pnpm/pnpm/tree/main/packages/make-dedicated-lockfile

However, I would not recommend to use separate lockfiles for each project. This will mean that each project will have its own virtual store (node_modules/.pnpm), and installations will be much slower.

If the issue is with conflicts in the lockfile, then maybe there are other solutions to that.

**Reply by EscapeB** [reply]
Actually it's not just about conflict. 

There are hundreds of projects in our monorepo and they shared one single lockfile. We have  almost 200+ CI pipelines jobs to make sure every MR/PR should not break the e2e/unit test in every project and every project should build as expected.

Shared lockfile means that one project may "affected" by any MR/PR that changes the single shared lockfile, so we have to run all pipelines once a MR/PR changed the shared lockfile. 

The pipeline waiting time before every MR/PR get merged  is really intolerable. We want to make every project more "isolate" so that we can run a few of pipelines related with MR/PR.

Is it possible that workspace use separate lockfiles but shared one virtual store? 

I am not sure  whether the dependencies for each project are totally isolate if we shared one node_modules, especially for those packages that has phantom dependencies.

**Reply by EscapeB** [reply]
1. Try `make-dedicated-lockfile` and get new lockfile for one project.
2. Delete single shared lockfile and run `pnpm install` to generate totally new lockfile.

I got different lockfile struct in step1 and step2.

**Reply by lianghx-319** [reply]
> You can probably use this package to create the new lockfiles: https://github.com/pnpm/pnpm/tree/main/packages/make-dedicated-lockfile
> 
> However, I would not recommend to use separate lockfiles for each project. This will mean that each project will have its own virtual store (node_modules/.pnpm), and installations will be much slower.
> 
> If the issue is with conflicts in the lockfile, then maybe there are other solutions to that.

Can you dive into "maybe there are other solutions to that"?

**Comment by eric-burel** [comment]
Related issue in the Yarn ecosystem:
https://github.com/yarnpkg/berry/issues/1223#issuecomment-1177250693
and plugin I use for publication:
https://github.com/JanVoracek/yarn-plugin-entrypoint-lockfiles
and the postinstall script I use at Vulcan in a monorepo mixing packages and apps:
https://github.com/VulcanJS/vulcan-npm/blob/417393798f97da83e673cf220f6484d16ee0c6de/scripts/copy-yarn-lock.sh

I think this issue is legitimate and important. In the most useful scenarios (CI, Docker deployment and publication of read-only repositories, which is a common pattern), a transient lock file is enough. It doesn't have to stay in the repository, or it could have a non standard name, so that it's only used when intended to by explicitely providing the lockfile path in `pnpm install` command.

For now I will probably have to define a more advanced docker image that works at the mono repo level, loading the app code but also unpublished shared code. This way I can use the shared lock file. However it's a tad overkill as it will install all dependencies of the project.
